middleware/motore/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+#![feature(type_alias_impl_trait)]
+//! This crate provides the [sentinel](https://docs.rs/sentinel-core) middleware for [tower](https://docs.rs/tower).  
+//! See the [examples](https://github.com/sentinel-group/sentinel-rust/tree/main/middleware) for help.
+//! [http]: https://docs.rs/http
+
+use core::marker::PhantomData;
+use futures::Future;
+use motore::{layer::Layer, service::Service};
+use sentinel_core::EntryBuilder;
+
+/// The side where the middleware is deplyed.
+#[derive(Debug, Copy, Clone)]
+pub enum ServiceRole {
+    Server,
+    Client,
+}
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// It is used to extractor a resource name from requests for sentinel. If you the service request is [`http::Request`](https://docs.rs/http/latest/http/request/struct.Request.html),
+/// and you are using nightly toolchain, you don't need to provide a sentinel resource name extractor. The middleware will automatically extract the request uri.
+type Extractor<Cx, R> = fn(&Cx, &R) -> String;
+
+/// The fallback function when service is rejected by sentinel.
+type Fallback<S, Cx, R> =
+    fn(&Cx, &R, &sentinel_core::Error) -> Result<<S as Service<Cx, R>>::Response, BoxError>;
+
+/// The sentinel middleware service in motore.
+pub struct SentinelService<S, Cx, R>
+where
+    S: Service<Cx, R>,
```

Comment:
```suggestion
    S: Service<Cx, R> + Clone,
```
and we can derive `Clone` instead of the manually implemented `Clone` below?

Reply:
I remembered that the derivation failed even when `S` was `Clone`. I'll check it again.

Reply:
The `#[derive(Clone)]` macro does not work on the function pointer, eg., the `Extractor` and `Fallback` here. I met the same problem as here, https://stackoverflow.com/questions/31676504/how-to-clone-the-struct-include-rcfnt